In [43]:
db = """# 📋 Sehat Sathi — Comprehensive Project Report

---

## 1. Project Title

**Sehat Sathi: Your Health Assistant**

> *"Your Health Assistant, Anytime, Anywhere"*

---

## 2. Project Description

Sehat Sathi is a **mobile-friendly, multilingual web platform** that provides instant, accessible, and easy-to-understand healthcare guidance to users — primarily targeting **rural communities in India**. The platform integrates **Artificial Intelligence (AI) chatbot**, **Machine Learning (ML) disease prediction models**, **real-time WebRTC video consultations**, and **government healthcare scheme information** into a single unified application.

The application is designed to bridge the healthcare gap by putting a **virtual health assistant** in everyone's pocket, empowering individuals with the knowledge and tools to make informed health decisions.

---

## 3. Problem Statement

> **Millions of people in rural communities across India lack access to timely and reliable medical advice.**

This critical gap is caused by multiple barriers:

| Barrier | Impact |
|---------|--------|
| **Geographic Distance** | Rural areas lack nearby hospitals or clinics; patients must travel long distances |
| **Cost** | Healthcare costs are prohibitive for low-income families |
| **Lack of Information** | Limited awareness about diseases, government health schemes, and preventive care |
| **Language Barriers** | Most digital health tools are only available in English |
| **Doctor Shortage** | India has a doctor-to-patient ratio far below WHO recommendations in rural areas |

These barriers lead to **preventable health complications**, **delayed diagnoses**, and **perpetuated health inequality** across rural India.

---

## 4. Solution

Sehat Sathi addresses these problems through a multi-pronged approach:

```mermaid
graph TD
    A["Rural User"] --> B["Sehat Sathi Platform"]
    B --> C["🤖 AI Health Chatbot"]
    B --> D["🔬 ML Disease Prediction"]
    B --> E["📹 Video Consultation"]
    B --> F["🗺️ Hospital/Doctor Locator"]
    B --> G["🏛️ Government Schemes"]
    B --> H["📄 Insurance Management"]
    B --> I["📚 Health Guide"]
    
    C --> J["Instant Health Advice in 3 Languages"]
    D --> K["Symptom-Based + Skin Disease Detection"]
    E --> L["WebRTC Peer-to-Peer Calls"]
    F --> M["Google Maps with Verified Doctors"]
    G --> N["Eligibility-Based Scheme Filtering"]
    H --> O["Encrypted Document Storage"]
    I --> P["WHO-Based Health Education"]
```

**Key Differentiators:**
- **Multilingual support** (English, Hindi, Marathi) — making it accessible to non-English speakers
- **AI-powered** health guidance using Google's Gemini 2.5 Flash model
- **ML-based disease prediction** — both symptom-based and image-based (skin disease)
- **Live video consultations** with verified doctors via WebRTC
- **End-to-end encrypted** patient data with AES encryption
- **Government scheme discovery** with eligibility-based filtering

---

## 5. Features Overview

| # | Feature | Status | Description |
|---|---------|--------|-------------|
| 1 | AI Health Chatbot | ✅ Complete | 24/7 multilingual health guidance via Gemini AI |
| 2 | Symptom-Based Disease Prediction | ✅ Complete | ML model predicting top 3 diseases from symptoms |
| 3 | Skin Disease Detection | ✅ Complete | Deep learning image classification using MobileNetV2 |
| 4 | Live Video Consultation | ✅ Complete | WebRTC peer-to-peer video/voice calls |
| 5 | Doctor/Hospital Locator | ✅ Complete | Google Maps with verified doctor profiles |
| 6 | Government Scheme Discovery | ✅ Complete | Eligibility-based filtering of health schemes |
| 7 | Insurance Document Management | ✅ Complete | Secure upload/download/delete of documents |
| 8 | Health Guide | ✅ Complete | WHO-based health education content |
| 9 | Multilingual Interface | ✅ Complete | English, Hindi, Marathi with full UI translation |
| 10 | Authentication & Security | ✅ Complete | Firebase Auth + AES encryption + ReCAPTCHA |
| 11 | Doctor Portal | ✅ Complete | Registration, dashboard, schedule, appointments |
| 12 | Admin Panel | ✅ Complete | Doctor verification, user management |
| 13 | Prescription System | ✅ Complete | PDF generation, QR codes, digital prescriptions |
| 14 | Notifications | ✅ Complete | Real-time call/appointment notifications |
| 15 | Text-to-Speech | ✅ Complete | AI-generated voice responses using Gemini TTS |

---

## 6. Pages & Routes

### 6.1 Patient-Facing Pages

| Route | Page | Description |
|-------|------|-------------|
| `/` | Landing Page | Hero section, feature cards, CTA |
| `/login` | Authentication | Email/password login & signup |
| `/chatbot` | AI Chatbot | Full-screen chat interface with AI |
| `/health-guide` | Health Guide | Precautions, healthy habits, gov guidelines |
| `/government-schemes` | Government Schemes | Scheme browser + hospital locator + booking |
| `/insurance` | Insurance Documents | Upload, view, download, delete documents |
| `/map` | Doctor/Hospital Map | Google Maps + doctor listing + appointment booking |
| `/video-call` | Video Call | WebRTC video/voice consultation interface |
| `/about` | About Page | Project information |
| `/contact` | Contact Page | Contact form |
| `/profile` | User Profile | User profile management |
| `/patient/*` | Patient Dashboard | Patient-specific features |
| `/verify` | Verification | Account/identity verification |
| `/call` | Call Management | Call session handling |

### 6.2 Doctor-Facing Pages (Route Group: `(doctor)`)

| Route | Page | Description |
|-------|------|-------------|
| `/doctor` | Doctor Home | Entry point for doctors |
| `/doctor/login` | Doctor Login | Separate authentication for doctors |
| `/doctor/signup` | Doctor Registration | Multi-step registration with verification |
| `/doctor/onboarding` | Onboarding | Post-registration setup |
| `/doctor/dashboard` | Doctor Dashboard | Appointment management, stats, call controls |
| `/doctor/profile` | Doctor Profile | Edit profile, specialization, hospital info |
| `/doctor/schedule` | Schedule Manager | Set weekly availability schedules |
| `/doctor/pending` | Pending Verification | Waiting screen for unverified doctors |

### 6.3 API Routes

| Route | Method | Description |
|-------|--------|-------------|
| `/api/predict-disease` | POST | Proxy to symptom-based ML model (Flask) |
| `/api/skin-disease` | POST | Proxy to skin disease ML model (Flask) |
| `/api/generate-prescription-pdf` | POST | PDF generation for prescriptions |
| `/api/auth` | POST | Authentication helpers |
| `/api/admin` | POST | Admin operations |

---

## 7. System Architecture

```mermaid
graph TB
    subgraph "Frontend (Next.js 15)"
        FE["React 18 + TypeScript"]
        TC["Tailwind CSS + shadcn/ui"]
        RHF["React Hook Form + Zod"]
        GMAP["Google Maps API"]
        WEBRTC_FE["WebRTC Client"]
    end

    subgraph "AI Layer (Google Genkit)"
        GENKIT["Genkit Framework"]
        GEMINI["Gemini 2.5 Flash"]
        TTS["Gemini TTS (Text-to-Speech)"]
    end

    subgraph "ML Microservices (Python Flask)"
        SYMPTOM["Symptom Predictor<br/>Logistic Regression<br/>Port 5000"]
        SKIN["Skin Disease Detector<br/>MobileNetV2 (CNN)<br/>Port 5001"]
    end

    subgraph "Backend (Firebase)"
        AUTH["Firebase Auth"]
        FS["Cloud Firestore"]
        STORAGE["Cloud Storage"]
        FUNC["Cloud Functions"]
        APPCHECK["App Check (ReCAPTCHA v3)"]
    end

    subgraph "Deployment"
        NETLIFY["Netlify / Firebase Hosting"]
        APPHOST["Firebase App Hosting"]
    end

    FE --> GENKIT
    GENKIT --> GEMINI
    GENKIT --> TTS
    FE --> AUTH
    FE --> FS
    FE --> STORAGE
    FE --> GMAP
    FE --> SYMPTOM
    FE --> SKIN
    WEBRTC_FE --> FS
    FUNC --> FS
    FE --> NETLIFY
    FE --> APPHOST
```

### Architecture Decisions

| Decision | Rationale |
|----------|-----------|
| **Next.js App Router** | Server-side rendering, file-based routing, API routes |
| **Turbopack** | Fast development builds (`next dev --turbopack`) |
| **Firebase Firestore** | Real-time listeners for video call signaling, live appointment updates |
| **WebRTC + Firebase Signaling** | Peer-to-peer calls without a media server; Firebase handles offer/answer/ICE exchange |
| **Separate ML Microservices** | Decoupled Python Flask APIs for independent scaling and model updates |
| **AES Encryption (CryptoJS)** | Client-side encryption of sensitive patient data (phone, disease) before Firestore storage |
| **Google Genkit** | Type-safe AI flow definitions with structured input/output schemas via Zod |

---

## 8. Database Schema (Cloud Firestore)

### 8.1 Collections Overview

```mermaid
erDiagram
    USERS ||--o{ INSURANCE_DOCS : has
    DOCTORS ||--o{ APPOINTMENTS : receives
    PATIENTS ||--o{ APPOINTMENTS : creates
    APPOINTMENTS ||--o{ CALL_SESSIONS : triggers
    CALL_SESSIONS ||--o{ CANDIDATES : contains
    DOCTORS ||--o{ PRESCRIPTIONS : creates
    NOTIFICATIONS }|--|| USERS : targets

    USERS {
        string uid PK
        string email
        string displayName
        object profile
    }

    DOCTORS {
        string uid PK
        string fullName
        string email
        string specialization
        string hospitalName
        string hospitalAddress
        object hospitalLocation
        string state
        string district
        string village
        string contact
        string qualification
        number consultationFee
        number experience
        string bio
        string profilePicture
        boolean isVerified
        number verificationLevel
        object availability
        string type
    }

    PATIENTS {
        string uid PK
        string name
        number age
        string gender
    }

    APPOINTMENTS {
        string id PK
        string doctorId FK
        string patientId FK
        string doctorName
        object patientDetails
        string status
        string mode
        string timing
        timestamp scheduledTime
        timestamp createdAt
        timestamp completedAt
        string callStatus
    }

    CALL_SESSIONS {
        string sessionId PK
        string appointmentId FK
        string doctorId FK
        string patientId FK
        string mode
        string status
        object offer
        object answer
        boolean initiatorMuted
        boolean receiverMuted
    }

    CANDIDATES {
        string candidateId PK
        string candidate
        string sdpMid
        number sdpMLineIndex
        string senderId
        timestamp timestamp
    }

    PRESCRIPTIONS {
        string id PK
        string doctorId FK
        string patientId FK
        object details
    }

    INSURANCE_DOCS {
        string id PK
        string title
        string url
        string path
        timestamp uploadedAt
    }

    NOTIFICATIONS {
        string id PK
        string recipientId FK
        string type
        object data
    }
```

### 8.2 Security Rules Summary

| Collection | Read | Write | Delete |
|-----------|------|-------|--------|
| `users/{uid}` | Owner only | Owner only | Owner only |
| `users/{uid}/insuranceDocuments` | Owner only | Owner only | Owner only |
| `doctors/{uid}` | **Public** (anyone) | Owner or Admin | Owner or Admin |
| `patients/{uid}` | Owner or Admin | Owner or Admin | — |
| `appointments/{id}` | Participants only | Doctor (accept/reject), Patient (cancel) | — |
| `callSessions/{id}` | Participants only | Participants | Doctor only |
| `callSessions/{id}/candidates` | Authenticated | Sender only | Sender or session Doctor |
| `prescriptions/{id}` | Participants | Doctor (create), Admin (modify) | Admin only |
| `notifications/{id}` | Recipient only | Authenticated (create) | Recipient |

### 8.3 Storage Rules

| Path | Access |
|------|--------|
| `/doctors/{uid}/**` | Owner only |
| `/prescriptions/{appointmentId}/**` | Appointment participants only |
| All other paths | **Denied** |

---

## 9. Tech Stack

### 9.1 Frontend

| Technology | Version | Purpose |
|-----------|---------|---------|
| **Next.js** | 15.5.4 | React framework with App Router |
| **React** | 18.3.1 | UI library |
| **TypeScript** | 5.x | Type safety |
| **Tailwind CSS** | 3.4.1 | Utility-first CSS framework |
| **shadcn/ui** (Radix UI) | Latest | Pre-built accessible UI components |
| **Lucide React** | 0.475.0 | Icon library |
| **React Hook Form** | 7.54.2 | Form management |
| **Zod** | 3.24.2 | Schema validation |
| **Recharts** | 2.15.1 | Dashboard charts |
| **Google Maps API** | 2.20.7 | Interactive maps |
| **date-fns** | 3.6.0 | Date formatting |
| **Embla Carousel** | 8.6.0 | Carousel component |

### 9.2 Backend / Cloud

| Technology | Purpose |
|-----------|---------|
| **Firebase Authentication** | Email/Password auth, future OTP |
| **Cloud Firestore** | NoSQL document database |
| **Cloud Storage** | File uploads (insurance, prescriptions, profile images) |
| **Cloud Functions** | Server-side triggers (appointment status on call end) |
| **Firebase App Check** | ReCAPTCHA v3 bot protection |

### 9.3 AI & ML

| Technology | Version | Purpose |
|-----------|---------|---------|
| **Google Genkit** | 1.16.1 | AI orchestration framework |
| **Gemini 2.5 Flash** | Latest | LLM for health chatbot & guidance |
| **Gemini TTS** (gemini-2.5-flash-preview-tts) | Preview | Text-to-Speech audio generation |
| **Scikit-learn** (Python) | — | Logistic Regression for disease prediction |
| **TensorFlow/Keras** (Python) | — | MobileNetV2 for skin disease detection |
| **Flask** (Python) | — | ML model serving API |

### 9.4 Real-time Communication

| Technology | Purpose |
|-----------|---------|
| **WebRTC** | Peer-to-peer video/voice calls |
| **STUN Servers** (Google) | NAT traversal |
| **TURN Servers** (OpenRelay) | Relay fallback for restricted networks |
| **Firebase Firestore** | WebRTC signaling (offer/answer/ICE exchange) |

### 9.5 Security

| Technology | Purpose |
|-----------|---------|
| **CryptoJS (AES)** | Client-side encryption of sensitive data |
| **Firebase Security Rules** | Collection/document-level access control |
| **Firebase Storage Rules** | File-level access control |
| **ReCAPTCHA v3** | Bot protection via App Check |

### 9.6 DevOps & Tooling

| Technology | Purpose |
|-----------|---------|
| **Turbopack** | Fast development bundler |
| **Mocha** | Testing framework for Firestore rules |
| **Sharp / Jimp** | Image processing |
| **jsPDF** | Client-side PDF generation |
| **html2canvas** | HTML to image conversion |
| **QRCode** | QR code generation for prescriptions |
| **Netlify / Firebase Hosting** | Deployment platforms |

---

## 10. ML Models — Detailed Analysis

### 10.1 Model 1: Symptom-Based Disease Prediction

#### Overview

| Attribute | Value |
|-----------|-------|
| **Model Name** | Disease Prediction using Symptoms |
| **Algorithm** | Logistic Regression (Multi-class) |
| **Framework** | Scikit-learn (Python) |
| **Serving** | Flask REST API (Port 5000) |
| **Model File** | [model.pkl](file:///e:/sehat-sathi-main/Sehat-Sathi-main/ml-models/predict_disease/model.pkl) (47 KB) |
| **Encoder** | [label_encoder.pkl](file:///e:/sehat-sathi-main/Sehat-Sathi-main/ml-models/predict_disease/label_encoder.pkl) (904 bytes) |
| **Training Notebook** | [disease_prediction_using symptons.ipynb](file:///e:/sehat-sathi-main/Sehat-Sathi-main/ml-models/predict_disease/disease_prediction_using%20symptons.ipynb) |

#### Model Architecture

```mermaid
graph LR
    A["132 Binary<br/>Symptom Features"] --> B["Logistic Regression<br/>(Multi-class OVR)"]
    B --> C["Softmax<br/>Probabilities"]
    C --> D["Top 3 Predictions<br/>+ Confidence %"]
```

| Component | Details |
|-----------|---------|
| **Input** | 132 binary features (one-hot encoded symptoms) |
| **Output** | Probability distribution over all disease classes |
| **Feature Encoding** | Binary (1 = symptom present, 0 = absent) |
| **Label Encoding** | LabelEncoder mapping disease names → integers |
| **Prediction** | Returns top 3 diseases with probability percentages |

#### API Contract

**Request:**
```json
POST /predict
Content-Type: application/json

{
  "symptoms": ["itching", "skin_rash", "fatigue", "high_fever"]
}
```

**Response:**
```json
{
  "top_disease": "Fungal infection",
  "confidence": 92.5,
  "other_possible": [
    { "disease": "Allergy", "probability": 5.2 },
    { "disease": "Drug Reaction", "probability": 2.3 }
  ],
  "message": "This is AI-based prediction. Please consult a doctor."
}
```

#### Evaluation Metrics

> [!NOTE]
> Exact evaluation metrics are computed in the training notebook. The model uses `predict_proba()` for probability-based ranking, which is critical for healthcare applications where confidence scoring matters. Based on the Logistic Regression approach with 132 features:

| Metric | Expected Range |
|--------|---------------|
| **Accuracy** | ~90-95% (typical for symptom-disease datasets with clear symptom patterns) |
| **Method** | One-vs-Rest (OVR) multiclass classification |
| **Ranking** | Top-3 prediction with probability scores |
| **Threshold** | No binary threshold; returns ranked probabilities |

---

### 10.2 Model 2: Skin Disease Detection

#### Overview

| Attribute | Value |
|-----------|-------|
| **Model Name** | Skin Disease Detection |
| **Algorithm** | Convolutional Neural Network (CNN) — **MobileNetV2** |
| **Framework** | TensorFlow / Keras (Python) |
| **Serving** | Flask REST API (Port 5001) |
| **Model File** | [skin_model.h5](file:///e:/sehat-sathi-main/Sehat-Sathi-main/ml-models/skin_disease/skin_model.h5) (27 MB) |
| **Class Labels** | [class_names.pkl](file:///e:/sehat-sathi-main/Sehat-Sathi-main/ml-models/skin_disease/class_names.pkl) (131 bytes) |
| **Training Notebook** | [skin_disease_ditection_dl.ipynb](file:///e:/sehat-sathi-main/Sehat-Sathi-main/ml-models/skin_disease/skin_disease_ditection_dl.ipynb) |

#### Model Architecture

```mermaid
graph TD
    A["Input Image<br/>(224 × 224 × 3 RGB)"] --> B["Normalization<br/>(Pixel / 255.0)"]
    B --> C["MobileNetV2<br/>Base (Pre-trained on ImageNet)"]
    C --> D["Global Average Pooling"]
    D --> E["Dense Layers"]
    E --> F["Softmax Output<br/>(5 classes)"]
    F --> G["Predicted Disease<br/>+ Confidence %"]
```

| Component | Details |
|-----------|---------|
| **Base Model** | MobileNetV2 (pre-trained on ImageNet) — Transfer Learning |
| **Input Size** | 224 × 224 × 3 (RGB) |
| **Preprocessing** | Resize to 224×224, normalize to [0, 1] |
| **Output Classes** | 5 fixed skin disease categories |
| **Activation** | Softmax (final layer) |
| **Model Size** | ~27 MB (.h5 format) |

#### Disease Classes (5 Fixed Labels)

| # | Class Name | Description |
|---|-----------|-------------|
| 1 | **Acne** | Common skin condition with pimples and inflammation |
| 2 | **Fungal infection** | Skin infection caused by fungi |
| 3 | **Eczema** | Chronic inflammatory skin condition |
| 4 | **Psoriasis** | Autoimmune condition causing rapid skin cell growth |
| 5 | **Benign_tumors** | Non-cancerous skin growths |

#### API Contract

**Request:**
```
POST /predict
Content-Type: multipart/form-data

image: <uploaded image file>
```

**Response:**
```json
{
  "disease": "Acne",
  "confidence": 74.8
}
```

#### Evaluation Metrics

> [!NOTE]
> Detailed evaluation metrics are in the training notebook. As a MobileNetV2-based transfer learning model on a 5-class dataset:

| Metric | Expected Range |
|--------|---------------|
| **Architecture** | MobileNetV2 (depth-wise separable convolutions) |
| **Training Strategy** | Transfer Learning from ImageNet weights |
| **Input Resolution** | 224 × 224 pixels |
| **Parameters** | ~3.4M (MobileNetV2 base) + custom head |
| **Inference Time** | ~50-200ms per image |
| **Model Efficiency** | Optimized for mobile/edge deployment (lightweight CNN) |

---

## 11. AI Flows (Google Genkit) — Detailed

### 11.1 Personalized Health Guidance Flow

**File:** [personalized-health-guidance.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/ai/flows/personalized-health-guidance.ts)

| Attribute | Details |
|-----------|---------|
| **Model** | Gemini 2.5 Flash (`googleai/gemini-2.5-flash`) |
| **Input** | Symptoms (text), medical history (text), language (en/hi/mr), optional photo (base64 data URI) |
| **Output** | Tailored health advice (text) |
| **Safety** | System prompt explicitly states "You are not a doctor" and mandates doctor consultation recommendations |
| **Multimodal** | Supports text + image input (photo of symptom) |

**System Prompt Highlights:**
- Follows WHO guidelines
- Provides safe, general health guidance
- Always recommends consulting a certified doctor
- Responds in the user's selected language

### 11.2 Health Information Summarizer Flow

**File:** [summarize-health-info.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/ai/flows/summarize-health-info.ts)

| Attribute | Details |
|-----------|---------|
| **Model** | Gemini 2.5 Flash |
| **Input** | Large health information text |
| **Output** | Concise summary |
| **Use Case** | Summarizing long medical articles or health guidelines for easy understanding |

### 11.3 Text-to-Speech Flow

**File:** [text-to-speech.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/ai/flows/text-to-speech.ts)

| Attribute | Details |
|-----------|---------|
| **Model** | Gemini 2.5 Flash Preview TTS (`gemini-2.5-flash-preview-tts`) |
| **Voice** | Algenib (prebuilt voice) |
| **Output Format** | WAV audio (base64 data URI) |
| **Pipeline** | Text → Gemini TTS → PCM audio → WAV encoding → Base64 |
| **Use Case** | Reading out chatbot responses for users who cannot read |

---

## 12. Feature Deep Dives

### 12.1 🤖 AI Health Chatbot

**Component:** [ChatInterface.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/chatbot/ChatInterface.tsx) (14.7 KB)

| Aspect | Details |
|--------|---------|
| **AI Backend** | Google Genkit + Gemini 2.5 Flash |
| **Multilingual** | Responds in English, Hindi, or Marathi based on user's language selection |
| **Multimodal** | Accepts text queries + optional symptom photos |
| **TTS** | Chatbot responses can be played as audio via Gemini TTS |
| **Safety** | All responses include disclaimer to consult a real doctor |
| **Language Hook** | [use-chat-language.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-chat-language.ts) — persists language preference |

### 12.2 🔬 Symptom-Based Disease Prediction

**Component:** [PredictDiseaseTab.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/identify/PredictDiseaseTab.tsx) (14.2 KB)

| Aspect | Details |
|--------|---------|
| **Input** | User selects symptoms from a list of 132 symptoms (loaded from [symptoms.json](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/data/symptoms.json)) |
| **Model** | Logistic Regression served via Flask API |
| **Output** | Top 3 predicted diseases with confidence percentages |
| **Flow** | Frontend → Next.js API route (`/api/predict-disease`) → Flask server (port 5000) |
| **Feature List** | [features.json](file:///e:/sehat-sathi-main/Sehat-Sathi-main/ml-models/predict_disease/features.json) — 132 symptom features |

### 12.3 🧬 Skin Disease Detection

**Component:** [SkinDiseaseTab.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/identify/SkinDiseaseTab.tsx) (11 KB)

| Aspect | Details |
|--------|---------|
| **Input** | User uploads a photo of the affected skin area |
| **Model** | MobileNetV2 (CNN) served via Flask API |
| **Preprocessing** | Image resized to 224×224, normalized to [0,1] |
| **Output** | Predicted disease class + confidence percentage |
| **Flow** | Frontend → Next.js API route (`/api/skin-disease`) → Flask server (port 5001) |
| **Classes** | Acne, Fungal infection, Eczema, Psoriasis, Benign tumors |

### 12.4 📹 Live Video Consultation (WebRTC)

**Components:**
- [VideoCall.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/video/VideoCall.tsx) (15.7 KB) — UI component
- [webrtc.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/lib/webrtc.ts) — WebRTC service class
- [signaling.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/lib/signaling.ts) — Firebase signaling service
- [use-video-call.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-video-call.ts) — React hook orchestrating call lifecycle

| Aspect | Details |
|--------|---------|
| **Protocol** | WebRTC with ICE (STUN/TURN) |
| **Signaling** | Firebase Firestore real-time listeners |
| **Modes** | Video call, Voice-only call |
| **Controls** | Mute audio, toggle video, end call |
| **Roles** | Doctor = Initiator (creates offer), Patient = Responder (creates answer) |
| **ICE Servers** | Google STUN servers + OpenRelay TURN servers |
| **Status Sync** | Mute states synced in real-time via Firestore |
| **Cleanup** | Cloud Function (`onCallSessionDeleted`) auto-marks appointments as completed |

**Call Flow:**
```mermaid
sequenceDiagram
    participant Doctor
    participant Firestore
    participant Patient
    
    Doctor->>Firestore: Create callSession doc
    Doctor->>Doctor: getUserMedia() → local stream
    Doctor->>Doctor: Create RTCPeerConnection
    Doctor->>Firestore: Store SDP Offer
    
    Patient->>Firestore: Listen to callSession
    Firestore-->>Patient: Receive Offer
    Patient->>Patient: getUserMedia() → local stream
    Patient->>Patient: setRemoteDescription(offer)
    Patient->>Patient: createAnswer()
    Patient->>Firestore: Store SDP Answer
    
    Firestore-->>Doctor: Receive Answer
    Doctor->>Doctor: setRemoteDescription(answer)
    
    loop ICE Exchange
        Doctor->>Firestore: Send ICE candidate
        Firestore-->>Patient: Receive ICE candidate
        Patient->>Firestore: Send ICE candidate
        Firestore-->>Doctor: Receive ICE candidate
    end
    
    Note over Doctor,Patient: P2P Media Stream Established ✅
    
    Doctor->>Firestore: Delete callSession (end call)
    Firestore-->>Patient: Session deleted → end call
```

### 12.5 🗺️ Doctor & Hospital Locator

**Page:** [map/page.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/app/map/page.tsx) (39 KB — largest page in the project)

| Aspect | Details |
|--------|---------|
| **Map Provider** | Google Maps JavaScript API via `@react-google-maps/api` |
| **Data Sources** | Firestore (verified doctors) + static data (government hospitals) |
| **Filtering** | By state, district, village + free-text search |
| **Doctor Types** | Private Doctors, Government Hospitals |
| **Availability** | Real-time availability check based on weekly schedule |
| **Booking** | In-page appointment form with video/voice/in-person options |
| **Encryption** | Patient phone and disease info encrypted with AES before storing |
| **Double-Booking Prevention** | Checks existing appointments for same doctor/time slot |
| **Privacy Mode** | Option to hide patient name from doctor ("Private Patient") |
| **My Requests Tab** | View and manage existing appointment requests |

**Sub-Components:**
- [DoctorDetailCard.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/map/DoctorDetailCard.tsx) (18.6 KB) — Rich doctor profile card
- [GoogleMapEmbed.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/services/GoogleMapEmbed.tsx) (5.9 KB) — Reusable map component
- [MyRequestsTab.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/patient/MyRequestsTab.tsx) (19.6 KB) — Appointment request management

### 12.6 🏛️ Government Scheme Discovery

**Page:** [government-schemes/page.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/app/government-schemes/page.tsx) (14.7 KB)

| Aspect | Details |
|--------|---------|
| **Schemes** | 6 major government healthcare initiatives |
| **Filtering** | Age, gender, state, district-based eligibility filtering |
| **Hospital Locator** | Embedded Google Map showing government health centers |
| **Booking** | In-page appointment booking form for government hospitals |
| **Validation** | Zod schemas for filter and booking forms |
| **Schemes Shown** | Ayushman Bharat, National Health Mission, PMMVY, JSSK, RBSK, NTCP |

### 12.7 📄 Insurance Document Management

**Page:** [insurance/page.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/app/insurance/page.tsx) (10.8 KB)

| Aspect | Details |
|--------|---------|
| **Upload** | Firebase Cloud Storage with progress indicator |
| **Formats** | PDF, DOC, DOCX, JPG, JPEG, PNG (max 10 MB) |
| **Storage** | `insuranceDocuments/{userId}/{timestamp}_{filename}` |
| **Operations** | Upload, view, download, delete |
| **Access Control** | Only authenticated owner can access their documents |
| **Real-time** | Firestore `onSnapshot` for live document list updates |
| **Authentication** | Redirects unauthenticated users to login |

### 12.8 📚 Health Guide

**Page:** [health-guide/page.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/app/health-guide/page.tsx) (6.9 KB)

| Aspect | Details |
|--------|---------|
| **Content** | WHO-based health education |
| **Sections** | Precautions, Healthy Habits, Government Guidelines |
| **Interaction** | Click-to-expand detail dialogs |
| **External Links** | WHO India, Ministry of Health & Family Welfare |
| **Multilingual** | All content translated via translations system |

### 12.9 💊 Prescription System

**Components:**
- [PrescriptionDialog.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/patient/PrescriptionDialog.tsx) (9.3 KB) — View/download prescriptions
- API route for PDF generation (`/api/generate-prescription-pdf`)

| Aspect | Details |
|--------|---------|
| **Creation** | Doctor creates prescriptions during/after consultations |
| **Format** | Digital view + downloadable PDF |
| **QR Code** | QR codes generated for prescription verification |
| **PDF Engine** | jsPDF + html2canvas for client-side PDF generation |
| **Storage** | Firestore (`prescriptions` collection) |

### 12.10 🔔 Notifications

**Component:** [CallNotification.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/components/notifications/CallNotification.tsx) (6.3 KB)

| Aspect | Details |
|--------|---------|
| **Types** | Incoming call notifications, appointment updates |
| **Delivery** | Real-time via Firestore `onSnapshot` listeners |
| **Collection** | `notifications` with `recipientId` for access control |

### 12.11 🔐 Authentication & Security

| Component | Implementation |
|-----------|---------------|
| **Patient Auth** | Email/Password via Firebase Auth ([login/page.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/app/login/page.tsx)) |
| **Doctor Auth** | Separate login flow with verification workflow |
| **Data Encryption** | AES encryption via CryptoJS ([encryption.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/lib/encryption.ts)) |
| **Bot Protection** | ReCAPTCHA v3 via Firebase App Check |
| **Auth Provider** | [FirebaseProvider](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-firebase.tsx) wraps entire app with auth context |
| **User Hook** | [use-user.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-user.ts) for accessing current user state |

### 12.12 🌐 Multilingual System

| Aspect | Details |
|--------|---------|
| **Languages** | English (`en`), Hindi (`hi`), Marathi (`mr`) |
| **Implementation** | Static translation dictionaries ([translations.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/lib/translations.ts) — **209 KB**, the largest source file) |
| **Locale Files** | [identify.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/locales/identify.ts) (8.8 KB) — disease identification translations |
| **Translation Hook** | [use-chat-language.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-chat-language.ts) — manages language selection |
| **Coverage** | Every UI element, form label, error message, and content section |

### 12.13 👨‍⚕️ Doctor Portal

| Feature | Implementation |
|---------|---------------|
| **Registration** | Multi-step signup with hospital info, specialization, location |
| **Verification** | 3-level verification system (admin-controlled) |
| **Dashboard** | Appointment overview, stats cards, quick actions |
| **Schedule** | Weekly availability editor (per day: start/end time, open/closed) |
| **Profile** | Edit personal info, hospital details, profile picture |
| **Appointments** | Accept/reject requests, start video calls, issue prescriptions |
| **Layout** | Custom doctor layout with separate navigation ([doctor/layout.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/app/(doctor)/doctor/layout.tsx)) |

---

## 13. Cloud Functions

**File:** [functions/index.js](file:///e:/sehat-sathi-main/Sehat-Sathi-main/functions/index.js)

| Function | Trigger | Purpose |
|----------|---------|---------|
| `onCallSessionDeleted` | Firestore `onDelete` on `callSessions/{sessionId}` | Automatically marks the linked appointment as `completed` with `callStatus: 'ended'` when a call session document is deleted |

---

## 14. Custom React Hooks

| Hook | File | Purpose |
|------|------|---------|
| `useFirebase` | [use-firebase.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-firebase.tsx) | Firebase context provider (app, auth, db, storage) |
| `useAuth` | use-firebase.tsx | Access Firebase Auth instance |
| `useFirestore` | use-firebase.tsx | Access Firestore instance |
| `useStorage` | use-firebase.tsx | Access Cloud Storage instance |
| `useCollection` | [use-firestore.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-firestore.tsx) | Real-time Firestore collection listener |
| `useDoc` | use-firestore.tsx | Real-time Firestore document listener |
| `useUser` | [use-user.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-user.ts) | Current authenticated user state |
| `useVideoCall` | [use-video-call.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-video-call.ts) | Complete video call lifecycle management |
| `useChatLanguage` | [use-chat-language.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-chat-language.ts) | Language selection and persistence |
| `useToast` | [use-toast.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-toast.ts) | Toast notification system |
| `useMobile` | [use-mobile.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/hooks/use-mobile.tsx) | Mobile device detection |

---

## 15. UI Component Library (shadcn/ui)

The project uses **36 shadcn/ui components** built on Radix UI primitives:

````carousel
### Form & Input Components
- Accordion, Alert Dialog, Button, Calendar, Carousel
- Checkbox, Collapsible, Dialog, Dropdown Menu
- File Upload (custom), Form, Input, Label
- Menubar, Popover, Progress, Radio Group
<!-- slide -->
### Display & Layout Components  
- Alert, Avatar, Badge, Card, Chart (Recharts)
- Scroll Area, Select, Separator, Sheet
- Sidebar, Skeleton, Slider, Switch
- Table, Tabs, Textarea, Toast, Tooltip
````

---

## 16. Project Statistics

| Metric | Value |
|--------|-------|
| **Total Source Files** | ~80+ TypeScript/JavaScript files |
| **Lines of Code (Frontend)** | ~5,000+ lines (excluding node_modules) |
| **Largest Source File** | [translations.ts](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/lib/translations.ts) — 209 KB (multilingual dictionary) |
| **Largest Page** | [map/page.tsx](file:///e:/sehat-sathi-main/Sehat-Sathi-main/src/app/map/page.tsx) — 781 lines |
| **ML Model Files** | 2 models (~27.4 MB total) |
| **UI Components** | 36 shadcn/ui + 10+ custom components |
| **Supported Languages** | 3 (English, Hindi, Marathi) |
| **Firestore Collections** | 7 (users, doctors, patients, appointments, callSessions, prescriptions, notifications) |
| **API Routes** | 5 |
| **Custom Hooks** | 11 |
| **AI Flows** | 3 (health guidance, summarization, TTS) |

---

## 17. Deployment Configuration

| Platform | Config File | Details |
|----------|-------------|---------|
| **Firebase** | [firebase.json](file:///e:/sehat-sathi-main/Sehat-Sathi-main/firebase.json), [.firebaserc](file:///e:/sehat-sathi-main/Sehat-Sathi-main/.firebaserc) | Firestore, Storage, Functions, App Hosting |
| **Firebase App Hosting** | [apphosting.yaml](file:///e:/sehat-sathi-main/Sehat-Sathi-main/apphosting.yaml) | Next.js hosting configuration |
| **Netlify** | [netlify.toml](file:///e:/sehat-sathi-main/Sehat-Sathi-main/netlify.toml) | Alternative deployment target |
| **Environment** | `.env.local` | Firebase keys, API keys, ML URLs, crypto secrets |

---

## 18. Disclaimer

> [!CAUTION]
> This application is meant to provide **general health guidance and information only**. It is **not a substitute for professional medical advice, diagnosis, or treatment**. In case of an emergency, please contact a certified doctor or nearby hospital immediately.

---

*Report generated on June 3, 2026 by scanning the complete Sehat Sathi codebase.*

---

## 19. Current Implementation Addendum (September 2026)

The current implementation extends the original report in several important ways:

- The landing page now explains the patient journey, doctor journey, community resources, verification workflow, platform pages, privacy practices, and AI safety limits.
- The patient identification page has separate symptom prediction, skin-disease screening, and skin-cancer screening flows.
- The active skin-cancer model is a PyTorch EfficientNet-B3 checkpoint at `ml-models/skin_disease/trained_models_v3-final/efficientnet_b3_FINAL.pth`. It predicts `MEL`, `NV`, `BCC`, `AKIEC`, `BKL`, `DF`, and `VASC`.
- Skin-cancer education is localized in English, Hindi, and Marathi and includes causes, precautions, possible effects, seriousness, and doctor recommendations.
- Symptom and skin-disease results also expose structured patient-friendly information instead of only a short description.
- Prediction endpoints use an 85% confidence gate. Low-confidence results return a safe inability-to-identify response rather than a weak disease label.
- The unified ML backend is a separate Flask service and can be deployed independently from the Next.js frontend. Its current runtime includes the symptom model, TensorFlow skin-disease model, PyTorch skin-cancer model, and face-verification services.
- The active large model is tracked with Git LFS. Unused model checkpoints and generated Python caches should not be committed.

For the operational route map, local commands, ML contracts, verification details, and deployment instructions, use `Sehat-Sathi-main/docs/PROJECT_GUIDE.md` as the current contributor reference.
"""

In [44]:
import tensorflow as tf 
from tensorflow.keras.preprocessing.text import Tokenizer

In [45]:
tokenizer = Tokenizer()

In [46]:
tokenizer.fit_on_texts([db])

In [47]:
tokenizer.word_index

{'sehat': 1,
 'sathi': 2,
 'main': 3,
 'file': 4,
 'doctor': 5,
 'e': 6,
 'disease': 7,
 'string': 8,
 'tsx': 9,
 'health': 10,
 'src': 11,
 'skin': 12,
 'model': 13,
 'firestore': 14,
 'ts': 15,
 'to': 16,
 'patient': 17,
 'firebase': 18,
 'use': 19,
 'for': 20,
 '—': 21,
 '2': 22,
 'with': 23,
 '3': 24,
 'and': 25,
 'the': 26,
 'page': 27,
 '5': 28,
 'call': 29,
 'api': 30,
 'a': 31,
 'ml': 32,
 'based': 33,
 'ai': 34,
 'video': 35,
 'details': 36,
 'only': 37,
 'symptom': 38,
 '1': 39,
 'in': 40,
 'gemini': 41,
 'app': 42,
 'components': 43,
 'complete': 44,
 'kb': 45,
 'webrtc': 46,
 'government': 47,
 '✅': 48,
 '12': 49,
 'prediction': 50,
 'google': 51,
 'time': 52,
 'language': 53,
 'storage': 54,
 'via': 55,
 '6': 56,
 'ui': 57,
 'text': 58,
 'hooks': 59,
 'of': 60,
 'user': 61,
 'doctors': 62,
 'end': 63,
 'auth': 64,
 'prescriptions': 65,
 'appointment': 66,
 'br': 67,
 'is': 68,
 'models': 69,
 'real': 70,
 'b': 71,
 '9': 72,
 'verification': 73,
 'tts': 74,
 'flask': 75,
 '

In [48]:
# find sentence in dataset
input_sequences = []
for sentence in db.split('\n'):
    tokenized_sentence = tokenizer.texts_to_sequences([sentence])[0]
    
    for i in range(1, len(tokenized_sentence)):
        input_sequences.append (tokenized_sentence[:i+1])

In [49]:
input_sequences

[[602, 1],
 [602, 1, 2],
 [602, 1, 2, 21],
 [602, 1, 2, 21, 603],
 [602, 1, 2, 21, 603, 120],
 [602, 1, 2, 21, 603, 120, 296],
 [39, 120],
 [39, 120, 388],
 [1, 2],
 [1, 2, 389],
 [1, 2, 389, 10],
 [1, 2, 389, 10, 297],
 [389, 10],
 [389, 10, 297],
 [389, 10, 297, 604],
 [389, 10, 297, 604, 605],
 [22, 120],
 [22, 120, 143],
 [1, 2],
 [1, 2, 68],
 [1, 2, 68, 31],
 [1, 2, 68, 31, 196],
 [1, 2, 68, 31, 196, 390],
 [1, 2, 68, 31, 196, 390, 121],
 [1, 2, 68, 31, 196, 390, 121, 606],
 [1, 2, 68, 31, 196, 390, 121, 606, 197],
 [1, 2, 68, 31, 196, 390, 121, 606, 197, 607],
 [1, 2, 68, 31, 196, 390, 121, 606, 197, 607, 391],
 [1, 2, 68, 31, 196, 390, 121, 606, 197, 607, 391, 392],
 [1, 2, 68, 31, 196, 390, 121, 606, 197, 607, 391, 392, 298],
 [1, 2, 68, 31, 196, 390, 121, 606, 197, 607, 391, 392, 298, 25],
 [1, 2, 68, 31, 196, 390, 121, 606, 197, 607, 391, 392, 298, 25, 393],
 [1, 2, 68, 31, 196, 390, 121, 606, 197, 607, 391, 392, 298, 25, 393, 16],
 [1, 2, 68, 31, 196, 390, 121, 606, 197, 607

In [50]:
#find max length sentence
max_length = max([len(x) for x in input_sequences])

In [51]:
#Use padding to match all input size to 56
from tensorflow.keras.preprocessing.sequence import pad_sequences
padded_input_sequences = pad_sequences(input_sequences, maxlen = max_length, padding = 'pre')

In [52]:
padded_input_sequences

array([[   0,    0,    0, ...,    0,  602,    1],
       [   0,    0,    0, ...,  602,    1,    2],
       [   0,    0,    0, ...,    1,    2,   21],
       ...,
       [   0,    0,    0, ...,  245,   26,  195],
       [   0,    0,    0, ...,   26,  195, 1191],
       [   0,    0,    0, ...,  195, 1191, 1192]],
      shape=(4052, 56), dtype=int32)

In [53]:
# for every sequence remove last word, which is seprate input and output
x = padded_input_sequences[:,:-1]
y = padded_input_sequences[:,-1]

In [54]:
x.shape

(4052, 55)

In [55]:
y.shape

(4052,)

In [56]:
len(tokenizer.word_index)

1192

In [57]:
from tensorflow.keras.utils import to_categorical
y = to_categorical(y, num_classes = 1193)

In [58]:
y.shape

(4052, 1193)

In [59]:
y

array([[0., 1., 0., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 0., 1.]], shape=(4052, 1193))

In [62]:
from tensorflow.keras import Sequential, Input
from tensorflow.keras.layers import Embedding, LSTM, Dense

model = Sequential([
    Input(shape=(55,)),
    Embedding(input_dim=1193, output_dim=100),
    LSTM(150),
    Dense(1193, activation="softmax")
])

model.compile(
    loss="categorical_crossentropy",
    optimizer="adam",
    metrics=["accuracy"]
)

model.summary()

Model: "sequential_5"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ embedding_5 (Embedding)         │ (None, 55, 100)        │       119,300 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ lstm_5 (LSTM)                   │ (None, 150)            │       150,600 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1193)           │       180,143 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 450,043 (1.72 MB)

 Trainable params: 450,043 (1.72 MB)

 Non-trainable params: 0 (0.00 B)

In [63]:
x

array([[   0,    0,    0, ...,    0,    0,  602],
       [   0,    0,    0, ...,    0,  602,    1],
       [   0,    0,    0, ...,  602,    1,    2],
       ...,
       [   0,    0,    0, ..., 1190,  245,   26],
       [   0,    0,    0, ...,  245,   26,  195],
       [   0,    0,    0, ...,   26,  195, 1191]],
      shape=(4052, 55), dtype=int32)

In [64]:
model.fit(x,y, epochs=100)

Epoch 1/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 11s 58ms/step - accuracy: 0.0271 - loss: 6.5899
Epoch 2/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 53ms/step - accuracy: 0.0592 - loss: 5.9915
Epoch 3/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 55ms/step - accuracy: 0.0992 - loss: 5.7285
Epoch 4/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 53ms/step - accuracy: 0.1185 - loss: 5.4890
Epoch 5/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 54ms/step - accuracy: 0.1422 - loss: 5.2398
Epoch 6/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 54ms/step - accuracy: 0.1599 - loss: 4.9849
Epoch 7/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 53ms/step - accuracy: 0.1829 - loss: 4.7268
Epoch 8/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 55ms/step - accuracy: 0.2076 - loss: 4.4633
Epoch 9/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 52ms/step - accuracy: 0.2241 - loss: 4.2004
Epoch 10/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 55ms/step - accuracy: 0.2478 - loss: 3.9370
Epoch 11/100
127/127 ━━━━━━━━━━━━━━━━━━━━ 7s 52ms/step - accuracy: 0.2786 - loss: 3.6778
Epoch 12/100
127/127 ━━━━━━━━

In [70]:
import numpy as np

In [81]:
text = "sehat sathi"

for i in range(10):
    #tokenize
    token_text = tokenizer.texts_to_sequences([text])[0]

    #padding
    padded_token_text = pad_sequences([token_text], maxlen=55, padding='pre')

    #predict
    pos = np.argmax(model.predict(padded_token_text))

    for word, index in tokenizer.word_index.items():
        if index == pos:
            text = text+ " "+ word
            print(text)

1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 58ms/step
sehat sathi is
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 65ms/step
sehat sathi is a
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 62ms/step
sehat sathi is a mobile
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 62ms/step
sehat sathi is a mobile friendly
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 78ms/step
sehat sathi is a mobile friendly multilingual
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step
sehat sathi is a mobile friendly multilingual web
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 60ms/step
sehat sathi is a mobile friendly multilingual web platform
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 54ms/step
sehat sathi is a mobile friendly multilingual web platform that
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 72ms/step
sehat sathi is a mobile friendly multilingual web platform that provides
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 86ms/step
sehat sathi is a mobile friendly multilingual web platform that provides instant


# LSTM Text Prediction – Practice Report

## 1. Dataset

For this practice, textual data was prepared using details from the **Sehat Sathi project document**. The purpose was to understand the complete preprocessing pipeline and implementation of an LSTM-based text prediction model.

## 2. Data Preprocessing

The text data was processed through the following steps:

1. **Sentence Extraction:** Relevant sentences were extracted from the project document.
2. **Tokenization:** The sentences were converted into numerical word tokens using a tokenizer.
3. **Sequence Creation:** Numerical word sequences were generated from the tokenized text.
4. **Padding:** The sequences were padded to a fixed length of **56 tokens**.
5. **Input and Output Separation:** The last token of each sequence was separated as the target output, while the first **55 tokens** were used as input.
6. **One-Hot Encoding:** The output word was converted into a one-hot encoded vector of **1193 classes**.

The resulting dataset contained **4,052 training sequences**, with an input shape of **(4052, 55)**.

## 3. LSTM Model Architecture

The model was implemented using TensorFlow/Keras and consisted of three main layers:

| Layer           | Output Shape    |  Parameters |
| --------------- | --------------- | ----------: |
| Embedding       | (None, 55, 100) |     119,300 |
| LSTM            | (None, 150)     |     150,600 |
| Dense (Softmax) | (None, 1193)    |     180,143 |
| **Total**       |                 | **450,043** |

The **Embedding layer** converts word indices into 100-dimensional word representations. The **LSTM layer** contains 150 units and learns sequential relationships between words. The final **Dense layer** uses softmax activation to predict the next word among 1,193 possible vocabulary classes.

## 4. Training

The model was compiled using:

* **Loss:** Categorical Crossentropy
* **Optimizer:** Adam
* **Metric:** Accuracy
* **Epochs:** 100

After training for 100 epochs, the model achieved approximately **90% training accuracy**.

## 5. Text Prediction

After training, the model was tested by providing a sequence of words as input. The input sequence was processed using the same tokenizer and padding procedure used during training. The trained LSTM then predicted the probability of the next word, and the word with the highest predicted probability was selected as the output.

## 6. Conclusion

This practice successfully demonstrated the complete workflow for **LSTM-based text prediction**, starting from raw project-document text and progressing through sentence extraction, tokenization, numerical encoding, sequence generation, padding, input/output preparation, model training, and next-word prediction. The model achieved approximately **90% accuracy on the training dataset**, demonstrating that the LSTM was able to learn word-sequence patterns present in the prepared Sehat Sathi text data.
